### PR TITLE
OOP: Frogger game

### PR DESCRIPTION
**Pull request by vv2529**
### Object-Oriented JavaScript

[Demo](https://vv2529.github.io/frontend-nanodegree-arcade-game) |
[Code base](https://github.com/vv2529/frontend-nanodegree-arcade-game)

The code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

File: submissions/vv2529/frogger/js/app.js [hunk 1]
```diff
@@ -0,0 +1,127 @@
+// A common class for Enemy and Player
+var Entity = function(x, y){
+	this.x = x;
+	this.y = y;
+}
+Entity.prototype.render = function(){
+	ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
```
Comment (OleksiyRudenko) [inline submissions/vv2529/frogger/js/app.js]: Doesn't seem `Entity` owns property `sprite`.

File: submissions/vv2529/frogger/js/app.js [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+	this.speed = speed;
+};
+Enemy.prototype = Object.create(Entity.prototype);
+// I refuse to write `Enemy.prototype` on every single line!
+Enemy.prototype = Object.assign(Enemy.prototype, {
+	constructor: Enemy,
+	sprite: 'images/enemy-bug.png',
+	width: 98,
+	height: 66,
+	update: function(dt){
```
Comment (OleksiyRudenko) [inline submissions/vv2529/frogger/js/app.js]: Please use `Class.prototype.methodName =` pattern to assign methods.
Also assign properties within `Enemy` function body.

File: submissions/vv2529/frogger/js/app.js [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+		this.x += this.speed * dt;
+
+		// Get an enemy back on track when it is gone
+		if(this.x > 800) this.restart();
+
+		// Check for collision
```
Comment (OleksiyRudenko) [inline submissions/vv2529/frogger/js/app.js]: Separation of Concerns principle is followed  (e.g. `update` method does only rendering and   doesn't **contain** any inline code to check e.g. collisions)

File: submissions/vv2529/frogger/js/app.js [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+		if((((this.x > player.x) && (this.x < player.x + player.width)) ||
```
Comment (OleksiyRudenko) [inline submissions/vv2529/frogger/js/app.js]: Classes should not refer to any global variables, like global variable `player`, which is an instance of `Player` class  (referring to global constants and globals provided by the gaming platform like `Resources` is OK);
Hint: pass `Player` instance as an argument to every enemy

File: submissions/vv2529/frogger/js/app.js [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+		if(newY < 0){
+			this.restart('win');
+		}
+	},
+	restart: function(msg){
+		console.log(msg+'!');
+		if(msg == 'win') getId('wins').textContent++;
+		else if(msg == 'lose') getId('losses').textContent++;
+
+		this.x = 200;
```
Comment (OleksiyRudenko) [inline submissions/vv2529/frogger/js/app.js]: A mysterious number. Please, demystify here and across the code using constants and calculations based on constants.

**Comment by OleksiyRudenko** [comment]
@vv2529 There are several ways to give collaborators and idea that you're ready for the next iteration:
- resolve conversations you believe you have elaborated on
- tell you're done explicitly

File: submissions/vv2529/frogger/js/app.js [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+	this.x = this.spawnX;
+	this.y = this.spawnY;
+}
+
+// Now instantiate your objects.
+const player = new Player(200, 400);
+const allEnemies = [
+	new Enemy(100, 65, 200, player),
+	new Enemy(0, 145, 100, player),
+	new Enemy(-100, 225, 400, player)
```
Comment (OleksiyRudenko) [inline submissions/vv2529/frogger/js/app.js]: A number of mysterious numbers. Please, pay more attention to comments.
>A mysterious number.
>> A mysterious number. Please, demystify here and **across the code** using constants and calculations based on constants.

Comment (vv2529) [reply submissions/vv2529/frogger/js/app.js]: These are just initial values for enemies that don't have much meaning, I can easily play around with numbers. But 65,145,225 should be replaced, I see.

Comment (OleksiyRudenko) [reply submissions/vv2529/frogger/js/app.js]: You may make those calculable e.g. `ROW_HEIGHT * 2 + ROW_Y_OFFSET` that would help a contributor to play with them as easy as you can.

**Comment by vv2529** [comment]
Got rid of as many numbers as possible.

**Comment by vv2529** [comment]
Thank you, I'll take that into account!

**Pull request merged**